File: src/systems/paths/SinePath.ts
import { PathData, PathType } from './PathData';

/**
 * Sinusoidal wave pattern - fish moves in a wave
 * Exact port of C# SinePath implementation
 */
export class SinePath {
  private readonly start: [number, number];
  private readonly end: [number, number];
  private readonly amplitude: number;
  private readonly frequency: number;
  private readonly fishId: number;
  private readonly seed: number;
  private readonly startTick: number;
  private readonly speed: number;

  constructor(
    fishId: number,
    seed: number,
    startTick: number,
    speed: number,
    start: [number, number],
    end: [number, number],
    amplitude: number,
    frequency: number
  ) {
    this.fishId = fishId;
    this.seed = seed;
    this.startTick = startTick;
    this.speed = speed;
    this.start = start;
    this.end = end;
    this.amplitude = amplitude;
    this.frequency = frequency;
  }

  /**
   * Get position along the sine wave path at normalized time t
   * Matches C# SinePath.GetPosition (lines 31-52)
   */
  getPosition(t: number): [number, number] {
    const baseX = this.start[0] + (this.end[0] - this.start[0]) * t;
    const baseY = this.start[1] + (this.end[1] - this.start[1]) * t;

    const dx = this.end[0] - this.start[0];
    const dy = this.end[1] - this.start[1];
    const length = Math.sqrt(dx * dx + dy * dy);

    const perpX = -dy / length;
    const perpY = dx / length;

    const offset = Math.sin(t * this.frequency * Math.PI * 2) * this.amplitude;

    const x = baseX + perpX * offset;
    const y = baseY + perpY * offset;

    return [x, y];
  }

  getPathType(): PathType {
    return PathType.Sine;
  }

  /**
   * Get serializable path data
   * Matches C# SinePath.GetPathData (lines 56-77)
   */
  getPathData(): PathData {
    const distance = Math.sqrt(
      Math.pow(this.end[0] - this.start[0], 2) +
      Math.pow(this.end[1] - this.start[1], 2)
    );

    return {
      fishId: this.fishId,
      pathType: PathType.Sine,
      seed: this.seed,
      startTick: this.startTick,
      speed: this.speed,
      controlPoints: [
        this.start,
        this.end,
        [this.amplitude, this.frequency]
      ],
      duration: distance / this.speed,
      loop: false
    };
  }
}

File: Server/Systems/Paths/SinePath.cs
using OceanKing.Server.Models;

namespace OceanKing.Server.Systems.Paths;

/// <summary>
/// Sinusoidal wave pattern - fish moves in a wave
/// </summary>
public class SinePath : IPath
{
    private readonly float[] _start;
    private readonly float[] _end;
    private readonly float[] _baseStart;
    private readonly float[] _baseEnd;
    private readonly float _amplitude;
    private readonly float _frequency;
    private readonly int _fishId;
    private readonly int _seed;
    private readonly int _startTick;
    private readonly float _speed;
    
    public SinePath(int fishId, int seed, int startTick, float speed, float[] start, float[] end, float[] baseStart, float[] baseEnd, float amplitude, float frequency)
    {
        _fishId = fishId;
        _seed = seed;
        _startTick = startTick;
        _speed = speed;
        _start = start;
        _end = end;
        _baseStart = baseStart;
        _baseEnd = baseEnd;
        _amplitude = amplitude;
        _frequency = frequency;
    }
    
    public float[] GetPosition(float t)
    {
        // Linear interpolation for base path using individual fish position
        float baseX = _start[0] + (_end[0] - _start[0]) * t;
        float baseY = _start[1] + (_end[1] - _start[1]) * t;
        
        // Calculate perpendicular direction for wave using SHARED base anchors
        // This ensures all fish in formation have synchronized wave phase
        float dx = _baseEnd[0] - _baseStart[0];
        float dy = _baseEnd[1] - _baseStart[1];
        float length = MathF.Sqrt(dx * dx + dy * dy);
        
        // Perpendicular unit vector
        float perpX = -dy / length;
        float perpY = dx / length;
        
        // Apply sine wave offset
        float offset = MathF.Sin(t * _frequency * MathF.PI * 2) * _amplitude;
        
        float x = baseX + perpX * offset;
        float y = baseY + perpY * offset;
        
        return new[] { x, y };
    }
    
    public PathType GetPathType() => PathType.Sine;
    
    public PathData GetPathData()
    {
        float distance = MathF.Sqrt(
            MathF.Pow(_end[0] - _start[0], 2) + 
            MathF.Pow(_end[1] - _start[1], 2)
        );
        
        return new PathData
        {
            FishId = _fishId,
            PathType = PathType.Sine,
            Seed = _seed,
            StartTick = _startTick,
            Speed = _speed,
            ControlPoints = new[] 
            { 
                _start, 
                _end, 
                new[] { _amplitude, _frequency } // Store wave parameters
            },
            Duration = distance / _speed,
            Loop = false  // Fish should exit screen smoothly, not loop back
        };
    }
}
